src/import/sst_service.rs

```
@@ -94,6 +96,100 @@ where
+        header.set_peer(context.take_peer());
+        header.set_region_id(region_id);
+        header.set_region_epoch(context.take_region_epoch());
+        let mut req = Request::default();
+        req.set_cmd_type(CmdType::Snap);
+        let mut cmd = RaftCmdRequest::default();
+        cmd.set_header(header.clone());
+        cmd.set_requests(vec![req].into());
+        let (cb, future) = paired_future_callback();
+
+        let router = self.router.clone();
+        let importer = self.importer.clone();
+        async move {
+            let mut resp = IngestResponse::default();
+            if let Err(e) = router.send_command(cmd, Callback::Read(cb)) {
+                resp.set_error(e.into());
+                return Ok(resp);
+            }
+
+            let mut res = future.await.map_err(Error::from)?;
+            fail_point!("import::sst_service::ingest");
+            let mut resp_header = res.response.take_header();
+            if resp_header.has_error() {
+                pb_error_inc(label, resp_header.get_error());
+                resp.set_error(resp_header.take_error());
+                return Ok(resp);
+            }
+
+            // Make ingest command.
+            let mut cmd = RaftCmdRequest::default();
+            cmd.set_header(header);
+            for sst in ssts.iter() {
+                let mut ingest = Request::default();
+                ingest.set_cmd_type(CmdType::IngestSst);
+                ingest.mut_ingest_sst().set_sst(sst.clone());
+                cmd.mut_requests().push(ingest);
+            }
+            cmd.mut_header().set_term(resp_header.get_current_term());
+
+            // Here we shall check whether the file has been ingested before. This operation
+            // must execute after geting a snapshot from raftstore to make sure that the
+            // current leader has applied to current term.
+            for sst in ssts.iter() {
+                if !importer.exist(&sst) {
+                    warn!(
```

Is it better to return an error than just ignore it?

---

No.... If I returned an error, lightning may retry again and again......

---

Why, if the sst file does not exist, lightning will just write another, this shouldn't cause dead lock